app/scripts/background.js

```diff
@@ -405,14 +405,73 @@ export async function loadStateFromPersistence() {
+  // we load a key from localStorage that we set later in this file containing
+  // the last known successfully ran migration version. If the migration
+  // version in versionedData differs from this stored value that is indicative
+  // of some sort of issue occuring.
+  const lastGoodMigration = global.localStorage.getItem('lastGoodMigration');
+  // localStorage is serialized to strings, so we need to convert it to a
+  // number.
+  const lastGoodMigrationNum =
+    typeof Number(lastGoodMigration) === 'number'
+      ? Number(lastGoodMigration)
+      : 0;
+  const currentVersionNumber = versionedData.meta?.version;
+
+  // A new install would have no lastGoodMigration, so we only want to check
+  // for existence and for a mismatch if the lastGoodMigrationNum is not the
+  // same. If this is the case we capture a message.
+  if (lastGoodMigration && lastGoodMigrationNum !== currentVersionNumber) {
+    sentry?.captureMessage('MetaMask - Migration Version Mismatch');
+  }
+
   // check if somehow state is empty
   // this should never happen but new error reporting suggests that it has
   // for a small number of users
   // https://github.com/metamask/metamask-extension/issues/3919
   if (versionedData && !versionedData.data) {
+    // Get possibly backed up vault from localstorage
+    const keyringVault = global.localStorage.getItem('metaMaskVault');
     // unable to recover, clear state
     versionedData = migrator.generateInitialState(firstTimeState);
-    sentry.captureMessage('MetaMask - Empty vault found - unable to recover');
+    if (global.localStorage.getItem('USER_OPTED_IN_TO_RESTORE') === 'true') {
+      // We persist the vault to localStorage as a failsafe in the event state
+      // corruption occurs.  When we get to this point we pull it out of
+      // localStorage and use it to at least allow the user to recover their
+      // accounts and backup their seed phrase. The rest of their settings will
+      // be wiped and returned to default. For the sake of transparency we will
+      // show an error screen to the user informing them what happened and that
+      // their settings are defaulted. The 'restoredFromBackup' flag is used to
+      // show this error screen in ui.js. Once they restart the app the flag is
+      // reverted to bypass the error screen.
+      if (keyringVault) {
+        versionedData.data.KeyringController = {
+          // Restore the vault from localstorage
+          vault: JSON.parse(keyringVault),
+          // Set a flag to indicate that the vault was restored from backup
+          restoredFromBackup: true,
+        };
+        // We have to set the completedOnboarding flag to true to avoid an
+        // Infinite routing loop where after unlocking they are redirected back
+        // to an unlock screen on the onboarding flow.
+        versionedData.data.OnboardingController = {
+          completedOnboarding: true,
```

Comment:
Hmm. It may be better to drop them into the last screen of onboarding instead, to let them modify privacy settings before the wallet starts. i.e. maybe we can fix this infinite redirect bug in the UI instead.

Or we could backup the privacy settings alongside the vault? It should be small, just a collection of boolean flags for the most part.

Tricky scenario.

Reply:
@Gudahtt I'll investigate what fixing the redirection would take. The tricky part, I think, will be that metametrics opt-in happens before wallet import/creation. So, if I can fix the redirection, perhaps I can set a flag that will skip just that part of onboarding.

Reply:
@Gudahtt @danjm and I walked through the onboarding stuff, the reason why without this flag I would get into the infinite loop is because I had an "unlocked" wallet without a "firstTimeFlowType" which resulted in some routing logic that would lock my wallet and drop me back at the password screen. 

We decided that adding a new 'firstTimeFlowType' of restore makes sense. I am going to implement that flow in other PRs where it will not be active. I imagine we'll want a E2E test to cover the restore process and the onboarding flow when in restore flow as part of this PR